### PR TITLE
Feature/clap v4 wip

### PR DESCRIPTION
**Pull request by hargut**
## Current Behavior

clap & clap_complete v3 are used

## Proposed Changes

upgrade clap & clap_complete to v4
- address v3 deprecation warnings
- upgrade to v4 & fixing compiler errors
- address v4 deprecation warnings

## Notes

this PR is not ready as the upgrade guide was only partially done

## Checks

- [ ] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository.

### CODE REVIEW

**Comment by hargut** [comment]
https://github.com/build-trust/ockam/issues/3506

**Comment by mrinalwadhwa** [comment]
@hargut awesome thank you for starting this. Let us know if you have questions along the way.

File: implementations/rust/ockam/ockam_command/src/lib.rs [hunk 1]
```diff
@@ -148,8 +148,12 @@ EXAMPLES:
     version,
     long_version = Version::long(),
     next_help_heading = "GLOBAL OPTIONS",
-    mut_arg("help", |a| a.help_heading("GLOBAL OPTIONS")),
-    mut_subcommand("help", |c| c.about("Print help information"))
+    // TODO: mut_arg & mut_subcommand for help
```
Comment (hargut) [inline implementations/rust/ockam/ockam_command/src/lib.rs]: @mrinalwadhwa 
The previously used mut_arg & mut_subcommand for help lead with the new version to runtime errors like:
```
thread 'main' panicked at 'Argument `help` is undefined'
thread 'main' panicked at 'Command `help` is undefined'
```
So it seems that `mut_` on "help" does not work any more as in the previous version. I've check the docs, but could not locate any indication that this was removed or disabled. Further the upgrade guide is not providing an information this is no longer possible.

https://docs.rs/clap/4.0.0-rc.2/clap/_derive/index.html
https://docs.rs/clap/4.0.0-rc.2/clap/struct.Command.html#method.mut_subcommand

Did I miss something?

It looks like "help" is no longer part of the clap::builder::command::Command subcommands Vec and therefore mut_subcommand fails on trying to mutate "help".
At top level with the mut_arg it seems to be a similar situation. Trying to add "help" leads to a duplicate error, and only works out after disable_help_flag = true.

Best regards,
Harald

Comment (hargut) [reply implementations/rust/ockam/ockam_command/src/lib.rs]: https://github.com/clap-rs/clap/pull/4056
https://github.com/clap-rs/clap/blob/master/CHANGELOG.md

Comment (mrinalwadhwa) [reply implementations/rust/ockam/ockam_command/src/lib.rs]: Thank you for researching that, makes sense. In the refactor to clap v4, I think we can disable the use of `mut_arg` and `mut_subcommand` ... get everything working in the new version. And then add the change suggested in https://github.com/clap-rs/clap/pull/4056

Comment (hargut) [reply implementations/rust/ockam/ockam_command/src/lib.rs]: I just had a look at the first test run results, but cannot understand what was going on:

The checkout action states:
https://github.com/build-trust/ockam/actions/runs/3108456670/jobs/5043177038#step:3:485
```
478:  HEAD is now at 0207e5a Merge 40df091a79a724158272f306e4a88f46279cfc98 into fe6e34b8f824ff3b56a3482a62c9ba8e652c7d64
```

One error from the build is:
https://github.com/build-trust/ockam/actions/runs/3108456670/jobs/5043177038#step:7:57
```
 56: error: Unknown `#[clap(long)]` attribute (`#[arg(long)] exists)
  --> implementations/rust/ockam/ockam_command/src/node/create.rs:87:12
   |
87 |     #[clap(long, hide = true)]
   |            ^^^^
```

https://github.com/build-trust/ockam/blob/40df091a79a724158272f306e4a88f46279cfc98/implementations/rust/ockam/ockam_command/src/node/create.rs#L87

The file does not contain any `#[clap(` at commit 40df091a79a724158272f306e4a88f46279cfc98.

:confused: :astonished:

Comment (hargut) [reply implementations/rust/ockam/ockam_command/src/lib.rs]: Ah, the part originates from a commit that was not there when the change was created:

https://github.com/build-trust/ockam/commit/fe6e34b8f824ff3b56a3482a62c9ba8e652c7d64#diff-afac108b9e4f2377c7932bef242708078b72d74836268ccc0ff4a09a24f67a53R87

**Pull request closed without merging**